### DISCUSSION TITLE
SD Booting Process Using YOCTO

### DISCUSSION THREAD

**Discussion by Shivaraj199** [polarfire-soc/polarfire-soc-documentation#307] [HSS and U-Boot] [answered]
Hello I am trying to boot the Linux kernel over SD card I am using Yocto to build the kernel Image and Rootfs,

My Sd card Partition

```
    Command (m for help): p
    Disk /dev/sdc: 7,51 GiB, 8053063680 bytes, 15728640 sectors
    Disk model: SD/MMC          
    Units: sectors of 1 * 512 = 512 bytes
    Sector size (logical/physical): 512 bytes / 512 bytes
    I/O size (minimum/optimal): 512 bytes / 512 bytes
    Disklabel type: gpt
    Disk identifier: 1D929B7F-E8AA-5B49-A7D3-2F140E3A1FFA
    
    Device     Start      End  Sectors   Size Type
    /dev/sdc1   2048     4000     1953 976,5K BIOS boot
    /dev/sdc2   4096     8000     3905   1,9M Linux filesystem
    /dev/sdc3   8192 15728606 15720415   7,5G Linux filesystem
    
    Command (m for help): 
```

My first partition Contains the HSS payloads which has U-Boot.bin
```
  sudo dd if=payload.bin of=/dev/sdc1 
  1187+1 records in
  1187+1 records out
  608064 bytes (608 kB, 594 KiB) copied, 0,0386021 s, 15,8 MB/s
```

I load the fitimage and DTB file generated in Yocto Build in the 2 partition which is of ext4 type

```
 sudo mount /dev/sdb2 /mnt/kernel
 sudo cp fitImage--5.15+git0+29c2077400-r0-icicle-kit-es-amp-20221122110305.bin /mnt/kernel
 sudo cp  mpfs-icicle-kit-context-a.dtb /mnt/kernel
  sudo umount  /mnt/kernel
```

And finally for my rootfs which is again ext4 partition for the .wic image generated by Yocto 

```
 sudo mount /dev/sdb3 /mnt/rootfs/
 cp mpfs-dev-cli-.wic /mnt/rootfs
 

```

I can successfully Boot the HSS payload and can enter the U-Boot Prompt. But from there I am not able to boot the kernel, Below is the image I can see from the SD CARD inserted In the Icicle kit

```
RISC-V # ext4ls mmc 0:2 /
<DIR>       4096 .
<DIR>       4096 ..
<DIR>      16384 lost+found
         4347924 fitImage--5.15+git0+29c2077400-r0-icicle-kit-es-amp-20221122110305.bin
           16747 mpfs-icicle-kit-context-a.dtb
RISC-V # ext4ls mmc 0:3 /
<DIR>       4096 .
<DIR>       4096 ..
<DIR>      16384 lost+found
               578813952 mpfs-dev-cli-.wic

```

Ther Error I get Failing to Initialze the SD card

```
[    5.207119] mpfs-generic-service mpfs-generic-service: Registered MPFS generic service
[    5.215392] of_cfs_init
[    5.218139] of_cfs_init: OK
[    5.221397] clk: Not disabling unused clocks
[    5.222912] sdhci-cdns 20008000.mmc: no tuning point found
[    5.222928] random: fast init done
[    5.234765] mmc0: tuning execution failed: -5
[    5.234791] mmc0: error -5 whilst initialising SD card
[    5.244560] random: crng init done
[    5.244680] Waiting for root device /dev/mmcblk0p3...
[    5.330797] mmc0: new ultra high speed SDR50 SDHC card at address 59b4
[    5.339536] mmcblk0: mmc0:59b4 SDU1  7.50 GiB 
[    5.420364] blk_update_request: I/O error, dev mmcblk0, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    5.432426] Buffer I/O error on dev mmcblk0, logical block 0, async page read
[    5.442322] blk_update_request: I/O error, dev mmcblk0, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    5.454291] Buffer I/O error on dev mmcblk0, logical block 0, async page read
[    5.461580]  mmcblk0: unable to read partition table
[    5.476953] VFS: Cannot open root device "mmcblk0p3" or unknown-block(179,3): error -6
[    5.485053] Please append a correct "root=" boot option; here are the available partitions:
[    5.493549] b300         7864320 mmcblk0 
[    5.493566]  driver: mmcblk
[    5.500480] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(179,3)
[    5.509007] CPU: 1 PID: 1 Comm: swapper/0 Not tainted 5.15.32-linux4microchip+fpga-2022.05 #1
[    5.517606] Hardware name: Microchip PolarFire-SoC Icicle Kit (DT)
[    5.523839] Call Trace:
[    5.526316] [<ffffffff80004820>] dump_backtrace+0x1c/0x24
[    5.531780] [<ffffffff804c4018>] show_stack+0x2c/0x38
[    5.536893] [<ffffffff804c6d1a>] dump_stack_lvl+0x40/0x58
[    5.542344] [<ffffffff804c6d46>] dump_stack+0x14/0x1c
[    5.547450] [<ffffffff804c4192>] panic+0xe8/0x292
[    5.552203] [<ffffffff8060136a>] mount_block_root+0x178/0x22a
[    5.558001] [<ffffffff8060157e>] mount_root+0x162/0x17c
[    5.563287] [<ffffffff806016da>] prepare_namespace+0x142/0x184
[    5.569182] [<ffffffff80600f82>] kernel_init_freeable+0x1ee/0x20a
[    5.575328] [<ffffffff804ca9a6>] kernel_init+0x1e/0x104
[    5.580613] [<ffffffff8000300c>] ret_from_exception+0x0/0xc
[    5.586248] SMP: stopping secondary CPUs
[    5.590238] ---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(179,3) ]---

```

Complete error log In the text file
[log.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/11191975/log.txt)
Is the partitions correct? Why the SD card Is not initialized?

Below is an Separate question
 I was trying to build the mpfs-initramfs-image provided in meta-yocto (No changes where made default setting). 
When i the Load the .wic file directly to eMMC using usbimager I can only see till U boot and process hangs when it enter kernel boot.

**Comment by rockoguido** [comment]
Hi.

Going to try to run through this myself, got a few general questions….


  1.  Using the latest master branch, which seems to correspond to 2023.02.1?
  2.  You are targeting the Icicle Kit?
  3.  What image are you building… core-image-minimal?

I’ll kick something off and see what happens…

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

[Text  Description automatically generated]
<http://www.microchip.com/emasters>Visit our new learning site at www.microchip.com/mu<http://www.microchip.com/mu>

From: Shivaraj ***@***.***>
Sent: Monday, April 10, 2023 9:21 AM
To: polarfire-soc/polarfire-soc-documentation ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [PolarFire-SoC] SD Booting Process Using YOCTO (Discussion #307)

EXTERNAL EMAIL: Do not click links or open attachments unless you know the content is safe

Hello I am trying to boot the Linux kernel over SD card I am using Yocto to build the kernel Image and Rootfs,

My Sd card Partition

    Command (m for help): p

    Disk /dev/sdc: 7,51 GiB, 8053063680 bytes, 15728640 sectors

    Disk model: SD/MMC

    Units: sectors of 1 * 512 = 512 bytes

    Sector size (logical/physical): 512 bytes / 512 bytes

    I/O size (minimum/optimal): 512 bytes / 512 bytes

    Disklabel type: gpt

    Disk identifier: 1D929B7F-E8AA-5B49-A7D3-2F140E3A1FFA



    Device     Start      End  Sectors   Size Type

    /dev/sdc1   2048     4000     1953 976,5K BIOS boot

    /dev/sdc2   4096     8000     3905   1,9M Linux filesystem

    /dev/sdc3   8192 15728606 15720415   7,5G Linux filesystem



    Command (m for help):

My first partition Contains the HSS payloads which has U-Boot.bin

  sudo dd if=payload.bin of=/dev/sdc1

  1187+1 records in

  1187+1 records out

  608064 bytes (608 kB, 594 KiB) copied, 0,0386021 s, 15,8 MB/s

I load the fitimage and DTB file generated in Yocto Build in the 2 partition which is of ext4 type

 sudo mount /dev/sdb2 /mnt/kernel

 sudo cp fitImage--5.15+git0+29c2077400-r0-icicle-kit-es-amp-20221122110305.bin /mnt/kernel

 sudo cp  mpfs-icicle-kit-context-a.dtb /mnt/kernel

  sudo umount  /mnt/kernel

And finally for my rootfs which is again ext4 partition for the .wic image generated by Yocto

 sudo mount /dev/sdb3 /mnt/rootfs/

 cp mpfs-dev-cli-.wic /mnt/rootfs





I can successfully Boot the HSS payload and can enter the U-Boot Prompt. But from there I am not able to boot the kernel, Below is the image I can see from the SD CARD inserted In the Icicle kit

RISC-V # ext4ls mmc 0:2 /

<DIR>       4096 .

<DIR>       4096 ..

<DIR>      16384 lost+found

         4347924 fitImage--5.15+git0+29c2077400-r0-icicle-kit-es-amp-20221122110305.bin

           16747 mpfs-icicle-kit-context-a.dtb

RISC-V # ext4ls mmc 0:3 /

<DIR>       4096 .

<DIR>       4096 ..

<DIR>      16384 lost+found

               0 mpfs-dev-cli-.wic



Ther Error I get Failing to Initialze the SD card

[    5.207119] mpfs-generic-service mpfs-generic-service: Registered MPFS generic service

[    5.215392] of_cfs_init

[    5.218139] of_cfs_init: OK

[    5.221397] clk: Not disabling unused clocks

[    5.222912] sdhci-cdns 20008000.mmc: no tuning point found

[    5.222928] random: fast init done

[    5.234765] mmc0: tuning execution failed: -5

[    5.234791] mmc0: error -5 whilst initialising SD card

[    5.244560] random: crng init done

[    5.244680] Waiting for root device /dev/mmcblk0p3...

[    5.330797] mmc0: new ultra high speed SDR50 SDHC card at address 59b4

[    5.339536] mmcblk0: mmc0:59b4 SDU1  7.50 GiB

[    5.420364] blk_update_request: I/O error, dev mmcblk0, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0

[    5.432426] Buffer I/O error on dev mmcblk0, logical block 0, async page read

[    5.442322] blk_update_request: I/O error, dev mmcblk0, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0

[    5.454291] Buffer I/O error on dev mmcblk0, logical block 0, async page read

[    5.461580]  mmcblk0: unable to read partition table

[    5.476953] VFS: Cannot open root device "mmcblk0p3" or unknown-block(179,3): error -6

[    5.485053] Please append a correct "root=" boot option; here are the available partitions:

[    5.493549] b300         7864320 mmcblk0

[    5.493566]  driver: mmcblk

[    5.500480] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(179,3)

[    5.509007] CPU: 1 PID: 1 Comm: swapper/0 Not tainted 5.15.32-linux4microchip+fpga-2022.05 #1

[    5.517606] Hardware name: Microchip PolarFire-SoC Icicle Kit (DT)

[    5.523839] Call Trace:

[    5.526316] [<ffffffff80004820>] dump_backtrace+0x1c/0x24

[    5.531780] [<ffffffff804c4018>] show_stack+0x2c/0x38

[    5.536893] [<ffffffff804c6d1a>] dump_stack_lvl+0x40/0x58

[    5.542344] [<ffffffff804c6d46>] dump_stack+0x14/0x1c

[    5.547450] [<ffffffff804c4192>] panic+0xe8/0x292

[    5.552203] [<ffffffff8060136a>] mount_block_root+0x178/0x22a

[    5.558001] [<ffffffff8060157e>] mount_root+0x162/0x17c

[    5.563287] [<ffffffff806016da>] prepare_namespace+0x142/0x184

[    5.569182] [<ffffffff80600f82>] kernel_init_freeable+0x1ee/0x20a

[    5.575328] [<ffffffff804ca9a6>] kernel_init+0x1e/0x104

[    5.580613] [<ffffffff8000300c>] ret_from_exception+0x0/0xc

[    5.586248] SMP: stopping secondary CPUs

[    5.590238] ---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(179,3) ]---



Complete error log In the text file
log.txt<https://github.com/polarfire-soc/polarfire-soc-documentation/files/11191975/log.txt>
Is the partitions correct? Why the SD card Is not initialized?

Below is an Separate question
I was trying to build the mpfs-initramfs-image provided in meta-yocto (No changes where made default setting).
When i the Load the .wic file directly to eMMC using usbimager I can only see till U boot and process hangs when it enter kernel boot.

**Reply by Shivaraj199** [reply]
Hello,
1.  Using the latest master branch, which seems to correspond to 2023.02.1?
>> No i tried with 2022.11, I can try with 2023.02.1 branch.
2.  You are targeting the Icicle Kit?
>>Yes I am targeting Icicle kit

3.  What image are you building… core-image-minimal?
 >> Yes I am trying to build core-image-minimal. and i get the .wic.gz file. When I upload the entire file there is no issues but when I try to build with individual artifacts like mentioned above it  throws me an error.
 The goal to run 2 applications which relies on Linux UIO driver on run of Individual harts that runs LINUX, using systemmd  commands. But I am unsuccessful with the kernel boot itself
 
 Thank you

**Comment by rockoguido** [comment]
As a follow-on I ran a test with the latest Master Branch, which looks to correspond to 2023.02.   following default instructions from https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp was able to build and deploy a core-image-minimal-dev without an issue and booted just fine.

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

[Text  Description automatically generated]
<http://www.microchip.com/emasters>Visit our new learning site at www.microchip.com/mu<http://www.microchip.com/mu>

**Comment by vfalanis** [comment, marked as answer]
Hi,

Is there a specific reason you want to update the components manually? This is doable. However it is recommended to use the WIC file if you are changing various components like payload, file system, kernel, etc. The WIC file should be very quicker to flash if using bmaptools, specially if you have a large file system.

Nevertheless, please find some comments below:

> And finally for my rootfs which is again ext4 partition for the .wic image generated by Yocto

This looks like the reason of the issue you are describing in this discussion. The WIC image contains a Linux image, so copying it to a partition inside the SD/mmc won't copy the file system. You would need to flash the core-image-minimal-dev-icicle-kit-es*.ext4 file located inside the build/tmp-glibc/deploy/images/icicle-kit-es/ directory of your Yocto build to the third partition of your SD/eMMC:

For example:
```
sudo dd if=core-image-minimal-dev-icicle-kit-es.ext4 of=/dev/sdX3
```
